File: process-elixir/lib/process.ex
require Logger
require Poison
require CSV

defmodule Process.CLI do

  def main(args) do
    case args do
      [field, folder] -> process(field, folder)
      _ -> Logger.error "Args are: <field name> <folder>"
    end
  end

  # # Benchmark
  # def main(_) do main_rec(100) end
  # def main_rec(0) do :done end
  # def main_rec(n) do process("Name", "../test_data/") end

  def process(field, folder) do
    values = for file <- glob(folder, ".json"),
        contents = File.read!(file),
        json = Poison.decode!(contents, as: %{}) do
      case json[field] do
        nil -> raise "Field is missing"
        x when not is_binary(x) -> raise "Field is not a string"
        text -> text
      end
    end

    frequencies = values
      |> Enum.filter(&(&1 != ""))
      |> Enum.reduce(%{}, fn(x, acc) -> Map.update(acc, x, 1, &(&1 + 1)) end)

    sorted_list = Map.to_list(frequencies)
      |> Enum.sort(&(elem(&1, 1) > elem(&2, 1)))
      |> Enum.map(&(Tuple.to_list(&1)))

    {:ok, outfile} = File.open("output.csv", [:write, :utf8])
    try do
      sorted_list |> CSV.encode |> Enum.each(&IO.write(outfile, &1))
    after
      File.close(outfile)
    end

    :ok
  end

  defp glob(folder, mask) do
    {:ok, regex} = Regex.compile(".*?" <> Regex.escape(mask))
    for filename <- File.ls!(folder),
        filename =~ regex,
        filepath = Path.join(folder, filename) do
      filepath
    end
  end
end

File: process-lua/process.lua
local lfs = require "lfs"
local globtopattern = require 'globtopattern'.globtopattern
local cjson = require "cjson"

function sort_dict_by_value(dict)
  local sorted = {}
  for k, v in pairs(dict) do
    table.insert(sorted, {k, v})
  end
  local function sortbyvalue(a, b)
    if a[2] == b[2] then
        return a[1] < b[1]
    else
        return a[2] > b[2]
    end
  end
  table.sort(sorted, sortbyvalue)
  return sorted
end

function escapeCSV(s)
  if string.find(s, '[,"]') then
    s = '"' .. string.gsub(s, '"', '""') .. '"'
  end
  return s
end

function process(field, folder)
  pattern = globtopattern("*.json")
  frequencies = {}
  for file in lfs.dir(folder) do
    if string.match(file, pattern) then
      local path = folder.."/"..file
      local f = io.open(path, "r")
      local contents = f:read("*all")
      f:close()
      local data = cjson.decode(contents)
      local value = data[field]
      if not value then
        error("Field is missing")
      end
      if type(value) ~= "string" then
        error("Field is not a string")
      end
      if value ~= "" then
        frequencies[value] = (frequencies[value] or 0) + 1
      end
    end
  end
  local sorted = sort_dict_by_value(frequencies)
  local of = io.open("output.csv", "w")
  for i, v in ipairs(sorted) do
    of:write(escapeCSV(v[1])..","..escapeCSV(v[2]).."\n")
  end
  of:close()
end
